src/app/ClusterObjectUtils.h

```diff
@@ -0,0 +1,312 @@
+
+        return count;
+    }
+
+    class Iterator {
+    public:
+        /*
+         * @brief
+         *
+         * Since neither the iterator constructor nor the operator overload permit sending errors back,
+         * encapsulate a tuple of the value and error that is then returned to the caller when decoding
+         */
+        struct ValuePair {
+            bool HasError() { return (status != CHIP_NO_ERROR); }
+            CHIP_ERROR status;
+            T value;
+        };
+
+        using iterator_category = std::input_iterator_tag;
+        using value_type = ValuePair;
+        using pointer = ValuePair*;
+        using reference = ValuePair&;
+
+        /*
+         * @brief
+         *
+         * Constructor
+         *
+         * reader       Pointer to a TLV reader whoose contents are copied into an internal
+         *              reader instance that is then used to track the iterator position 
+         *              within the backing buffer.
+         *
+         * signature    A unique signature that associated it with a given IteratableList instance
+         *              This permits actually comparing two iterators trivially to ensure they're
+         *              both logically iterating over the same list.
+         *
+         * isEnd        A flag that indicates if the iterator is pointing at the end of the TLV list.
+         *              This is preferable to actually iterating over the list till the end.
+         */
+        Iterator(TLV::TLVReader *reader, uintptr_t signature, bool isEnd = false) {
+            mSignature = signature;
+            mIsEnd = isEnd;
+            mValue.status = CHIP_NO_ERROR;
+            mDecoded = false;
+
+            if (!isEnd && reader) {
+                mReader.Init(*reader);
+                mValue.status = mReader.Next();
+            }
+        }
+
+        friend bool operator==(const Iterator& a, const Iterator& b) {
+            return ((a.mSignature == b.mSignature) && ((a.mIsEnd == b.mIsEnd)));
+        }
+
+        friend bool operator!=(const Iterator& a, const Iterator& b) {
+            return !((a.mSignature == b.mSignature) && ((a.mIsEnd == b.mIsEnd)));
+        }
+
+        /*
+         * @brief
+         *
+         * De-reference operator overload that returns a fully decoded list item
+         * at the current iterator position.
+         *
+         * Before accessing the value within the pair, the status should *always* be
+         * checked first.
+         *
+         * This can be called multiple times safely at the same iterator position.
+         *
+         */
+        ValuePair* operator->() {
+            if (mValue.status == CHIP_NO_ERROR && !mDecoded) {
+                mValue.status = Decode(mReader, mValue.value);
+                mDecoded = true;
+            }
+
+            return &mValue;
+        }
+
+        /*
+         * @brief
+         *
+         * Increments the iterator one item forward in the current list.
```

Comment:
Are there use cases for advancing the iterator but then not getting the value?  Because it would seem to me that if we just have a `Next` method with a `T&` outparam we won't need the temporary storage inside the iterator, etc....

Reply:
I had debated that initially, but having a T& outparam means you have to know the type of T at the call-site.

Right now, because I'm returning the decoded value as a return val, this permits the use of `auto`, which is I think, a worthwhile trade-off.

Reply:
Right, the tradeoff here is whether you decode directly into caller-provided memory or whether you decode into on-stack memory and then the caller has to copy out of there.

I guess if we're ok with the latter one option for an API is something like:
```
Optional<const T&> Next();
CHIP_ERROR GetLastError();
```
or
```
Optional<const T&> Next(CHIP_ERROR & error);
```
Or even:
```
const T& Next(CHIP_ERROR & error)
```

Though I guess we could keep this ValuePair thing as well; it's just a bit more mental overhead to using this class properly that people have to learn...

Reply:
Check out my most recent attempt in this space.